tasks/shell_crmsh/check-and-prepare-role-variables.yml

```diff
@@ -0,0 +1,135 @@
+  ansible.builtin.assert:
+    that: "'{{ item }}' in ansible_facts.services"
+    fail_msg: >-
+      The service '{{ item }}' was not found on this system. Ensure that this
+      service is available before running this role.
+    success_msg: >-
+      The service '{{ item }}' was discovered on this system.
+  loop:
+    - 'logd.service'
+
+- name: Discover cluster node names
+  ansible.builtin.set_fact:
+    __ha_cluster_node_name: "{{ ha_cluster.node_name | d(inventory_hostname) }}"
+
+- name: Collect cluster node names
+  ansible.builtin.set_fact:
+    __ha_cluster_all_node_names: "{{
+        ansible_play_hosts
+        | map('extract', hostvars, '__ha_cluster_node_name')
+        | list
+      }}"
+
+- name: Extract qdevice settings
+  ansible.builtin.set_fact:
+    __ha_cluster_qdevice_in_use: "{{ 'device' in ha_cluster_quorum }}"
+    __ha_cluster_qdevice_model: "{{ ha_cluster_quorum.device.model | d('') }}"
+    # This may set empty value, if it is not defined. Such value is not valid.
+    # It will be caught by crm validation before we try using it in the role.
+    __ha_cluster_qdevice_host: "{{
+        ha_cluster_quorum.device.model_options | d([])
+        | selectattr('name', 'match', '^host$')
+        | map(attribute='value') | list | last | d('')
+      }}"
+    __ha_cluster_qdevice_crm_address: "{{
+        ha_cluster_quorum.device.model_options | d([])
+        | selectattr('name', 'match', '^crm-address$')
+        | map(attribute='value') | list | last | d('')
+      }}"
+
+- name: Figure out if ATB needs to be enabled for SBD
+  ansible.builtin.set_fact:
+    # SBD needs ATB enabled if all of these are true:
+    # - sbd does not use devices (In check-and-prepare-role-variables.yml it
+    #   is verified that all nodes have the same number of devices defined.
+    #   Therefore it is enough to check devices of any single node.)
+    # - number of nodes is even
+    # - qdevice is not used
+    __ha_cluster_sbd_needs_atb: "{{
+        ha_cluster_sbd_enabled
+        and not ha_cluster.sbd_devices | d([])
+        and __ha_cluster_all_node_names | length is even
+        and not __ha_cluster_qdevice_in_use
+      }}"
+
+- name: Fail if SBD needs ATB enabled and the user configured ATB to be disabled
+  ansible.builtin.fail:
+    msg: Cannot set auto_tie_breaker to disabled when SBD needs it to be enabled
+  when:
+    - __ha_cluster_sbd_needs_atb | bool
+    - ha_cluster_quorum.options | d([])
+      | selectattr('name', 'match', '^auto_tie_breaker$')
+      | map(attribute='value') | select('in', ['0', 0]) | list | length > 0
+
+# Sets fence agent list from os_familu vars depending on user input.
+- name: Set __ha_cluster_fence_agent_packages_final fact
+  ansible.builtin.set_fact:
+    __ha_cluster_fence_agent_packages_final:
+      "{{ (ha_cluster_fence_agent_packages | length == 0) |
+        ternary(__ha_cluster_fence_agent_packages_default,
+        ha_cluster_fence_agent_packages) }}"
```

Comment:
This is exactly what my proposal does, without having to resort to `set_fact` and defining yet another variable.

Reply:
Oh, yes. Somehow, I missed your comment about this. @marcelmamula, can you try implementing @richm's proposal regarding this?

Reply:
defaults/main.yml does not contain any __private variables. Do you want to expose them there now?
Latest commit has option for users to define fence agents and then decide based on that decision, without having them to to go and replace this assignment in defaults.

Reply:
> defaults/main.yml does not contain any __private variables. Do you want to expose them there now?

Yes.  This is quite a common idiom in system roles.

> Latest commit has option for users to define fence agents and then decide based on that decision, without having them to to go and replace this assignment in defaults.

Not sure what you mean.  With my proposal:
* if the user defines `ha_cluster_fence_agent_packages` then the role will use that definition
* if the user does not define `ha_cluster_fence_agent_packages` then `ha_cluster_fence_agent_packages` will be set to the default value `__ha_cluster_fence_agent_packages_default` and the default value will be used

Reply:
See also https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_variables.html#ansible-variable-precedence

Reply:
@richm Both solution works, but I will update it so we can move past this.

Reply:
[citest]
Changes are pushed for

Reply:
@richm this can be marked as resolved?